atoma-state/src/handlers.rs

```diff
@@ -1767,3 +1908,1027 @@ mod utils {
+    /// // 1. Worst GPU (250W) is 50% between threshold (200W) and max (300W)
+    /// // 2. Score of 0.5 is then weighted by 0.2
+    /// ```
+    ///
+    /// # Notes
+    ///
+    /// - Returns 0.0 if the input slice is empty
+    /// - Uses partial_cmp for floating point comparisons, defaulting to 0.0 if comparison fails
+    /// - The linear scaling provides a smooth transition between optimal and critical power usage levels
+    /// - This score can be used as part of a larger node performance evaluation system
+    /// - Returns -1.0 if the input slice is empty (TODO: We will introduce other accelerators metrics in the future)
+    fn compute_gpu_power_score(
+        gpu_power_usages: &[f64],
+        gpu_power_threshold: f64,
+        gpu_power_max: f64,
+        gpu_power_weight: f64,
+    ) -> f64 {
+        if gpu_power_usages.is_empty() {
+            return -1.0;
+        }
+        // For each GPU, assign:
+        //   1.0 if power <= threshold,
+        //   0.0 if power >= max,
+        //   and linearly scaled in between.
+        // Finally, take the minimum score across all GPUs.
+        gpu_power_weight
+            * gpu_power_usages
+                .iter()
+                .map(|&power| {
+                    if power <= gpu_power_threshold {
+                        1.0
+                    } else if power >= gpu_power_max {
+                        0.0
+                    } else {
+                        1.0 - ((power - gpu_power_threshold)
+                            / (gpu_power_max - gpu_power_threshold))
+                    }
+                })
+                .min_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal))
+                .unwrap_or(0.0)
+    }
+
+    /// Calculates a normalized CPU usage score based on the percentage of CPU time used.
+    ///
+    /// This function computes a score based on the percentage of CPU time used, where lower usage
+    /// indicates better performance. The score is linearly scaled between 0.0 and 1.0, with 0.0
+    /// representing 100% CPU usage and 1.0 representing 0% CPU usage.
+    ///
+    /// # Arguments
+    ///
+    /// * `cpu_usage` - A f64 value representing the percentage of CPU time used (0.0 to 1.0)
+    /// * `cpu_score_weight` - A weight factor to scale the final score (typically between 0.0 and 1.0)
+    ///
+    /// # Returns
+    ///
+    /// Returns a f64 score between 0.0 and 1.0 where:
+    /// - 0.0 indicates 100% CPU usage
+    /// - 1.0 indicates 0% CPU usage
+    /// - Values between 0.0 and 1.0 indicate linear degradation based on CPU usage
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// let cpu_usage = 0.5;
+    /// let weight = 0.3;
+    ///
+    /// let score = compute_cpu_score(cpu_usage, weight);
+    /// // Returns 0.15 because:
+    /// ```
+    ///
+    /// # Notes
+    ///
+    /// - Returns 0.0 if the input is 100% CPU usage
+    fn compute_cpu_score(cpu_usage: f64, cpu_score_weight: f64) -> f64 {
```

Comment:
Hmmm I may be naive here but is the `cpu_usage` here considering all the cores a CPU might have? If not we should e.g.

```rust
fn compute_cpu_score(
    cpu_usage: f64,
    num_cores: i32,
    current_frequency: f64,
    max_frequency: f64,
    cpu_score_weight: f64
) -> f64 {
    let frequency_ratio = current_frequency / max_frequency;
    let core_utilization = cpu_usage / num_cores as f64;
    
    cpu_score_weight * (1.0 - core_utilization) * frequency_ratio
}
```

Reply:
I am not sure if I can collect the `current_frequency`. As far as I understand the frequency measures how many cycles a CPU can execute per second, and therefore this is a single value, per se. That said, it is a good number to capture, too.